data/service/api/v1/alerts.go

```diff
@@ -0,0 +1,78 @@
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/tidepool-org/platform/alerts"
+	"github.com/tidepool-org/platform/data/service"
+	"github.com/tidepool-org/platform/request"
+	platform "github.com/tidepool-org/platform/service"
+)
+
+func AlertsRoutes() []service.Route {
+	return []service.Route{
+		service.MakeRoute("POST", "/v1/alerts/:userID", Authenticate(UpsertAlert)),
+		service.MakeRoute("DELETE", "/v1/alerts/:userID", Authenticate(DeleteAlert)),
+	}
+}
+
+func DeleteAlert(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	r := sCtx.Request()
+	details := request.DetailsFromContext(ctx)
+	repo := sCtx.AlertsRepository()
+
+	if err := checkPermissions(details, r.PathParam("userID")); err != nil {
+		sCtx.RespondWithError(platform.ErrorUnauthorized())
+		return
+	}
+
+	cfg := &alerts.Config{}
+	if err := json.NewDecoder(r.Body).Decode(cfg); err != nil {
+		sCtx.RespondWithError(platform.ErrorJSONMalformed())
+	}
+	cfg.OwnerID = details.UserID()
+	if err := repo.Delete(ctx, cfg); err != nil {
+		sCtx.RespondWithError(platform.ErrorInternalServerFailure())
+		return
+	}
+}
+
+func UpsertAlert(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	r := sCtx.Request()
+	details := request.DetailsFromContext(ctx)
+	repo := sCtx.AlertsRepository()
+
+	if err := checkPermissions(details, r.PathParam("userID")); err != nil {
+		sCtx.RespondWithError(platform.ErrorUnauthorized())
+		return
+	}
+
+	cfg := &alerts.Config{}
+	if err := json.NewDecoder(r.Body).Decode(cfg); err != nil {
+		sCtx.RespondWithError(platform.ErrorJSONMalformed())
```

Comment:
This will return a "json is malformed" error when validation fails, which isn't correct.

Reply:
I don't see how that's not correct? Can you expand on that? Which error message would you use?

Reply:
Let's say an alert threshold that we have received is beyond some bounds we have defined. This doesn't mean the json is malformed. The returned response ought to indicate which field failed validation and why.

Reply:
I guess we just see that differently. I'm not opposed to semantic validation in there. It makes sense to me, as I consider that JSON to be malformed. It does not unmarshal into a valid go object (because the integer is out of range.) To me, that's malformed. While I can see your point, I think it's too fine a distinction to bother adding the distinction and not worth the added complexity.

But it's moot in this case, as you've shown me the existing validation package, and I'm glad to use that. :)

Reply:
Maybe part of it is that I don't see UnmarshalJSON as validating only the JSON, I see it as saying, I have type Foo, and UnmarshalJSON should take in some JSON, and produce a valid Foo. Therefore, if there's any incompatability in doing so, that indicates that the JSON provided is invalid. It might be "valid" JSON, but that doesn't make it a valid Foo.